### PR TITLE
[SYCL][COMPAT] Add math extend_v*4 to SYCLCompat

### PR DESCRIPTION
**Pull request by OuadiElfarouki**
This PR adds math `extend_v*4` operators (18 in total) along with unit-tests for signed and unsigned int32 cases.
*Some changes overlap with the previous `extend_v*2` PR #13953 and thus should be reviewed/merged first.

### CODE REVIEW

**Review by joeatodd — CHANGES_REQUESTED** [review]
Hey @OuadiElfarouki nice work! Just a couple comments just now - to replace `enable_if` w/ `static_assert`. Apologies again for sending you the wrong way!

I notice the test `math_extend_v.cpp` is *very* slow (284 seconds when I run it locally). I'll investigate that further.

File: sycl/include/syclcompat/math.hpp [hunk 1]
```diff
@@ -105,18 +128,100 @@ template <typename RetT, bool needSat, typename AT, typename BT, typename CT,
...
+  return ret;
+}
+
+template <typename RetT, bool NeedSat, bool NeedAdd, typename AT, typename BT,
+          typename BinaryOperation>
+inline constexpr std::enable_if_t<
+    std::is_integral_v<AT> && std::is_integral_v<BT> &&
+        std::is_integral_v<RetT> && sizeof(AT) == 4 && sizeof(BT) == 4 &&
+        sizeof(RetT) == 4,
+    RetT>
```
[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply sycl/include/syclcompat/math.hpp:173]: noted will do!

[the remark this answers is not in the document]
Comment (OuadiElfarouki) [reply sycl/include/syclcompat/math.hpp]: Done

**Review by Alcpz — APPROVED** [review]
LGTM

**Comment by OuadiElfarouki** [comment]
@intel/llvm-gatekeepers PR is ready for merge 🙏🏻 
Also CI `memory_management.cpp` failure is being tracked in Issue #14086 
Thanks!

**Comment by Alcpz** [comment]
Sorry @steffenlarsen , @joeatodd is off for a few days and I covered for him re-reviewing the changes. Is it ok or shall we still wait?

File: sycl/include/syclcompat/math.hpp [hunk 2]
```diff
@@ -151,6 +151,19 @@ template <typename T> sycl::vec<int32_t, 2> extractAndExtend2(T a) {
   return ret;
 }
 
+template <typename T> sycl::vec<int16_t, 4> extractAndExtend4(T a) {
```
Comment (yihanwg) [inline sycl/include/syclcompat/math.hpp:154]: I'd like to use snake_case rule here.

**Review by tangjj11 — APPROVED** [review]

File: sycl/include/syclcompat/math.hpp [hunk 3]
```diff
@@ -151,6 +151,19 @@ template <typename T> sycl::vec<int32_t, 2> extractAndExtend2(T a) {
...
+  sycl::vec<int16_t, 4> ret;
+  sycl::vec<T, 1> va{a};
+  using Tint =
+      typename std::conditional<std::is_signed_v<T>, int8_t, uint8_t>::type;
```
Comment (yihanwg) [inline sycl/include/syclcompat/math.hpp:158]:
```suggestion
std::conditional_t<std::is_signed_v<T>, int8_t, uint8_t>;
```

File: sycl/include/syclcompat/math.hpp [hunk 4]
```diff
@@ -151,6 +151,19 @@ template <typename T> sycl::vec<int32_t, 2> extractAndExtend2(T a) {
   return ret;
 }
 
+template <typename T> sycl::vec<int16_t, 4> extractAndExtend4(T a) {
+  sycl::vec<int16_t, 4> ret;
+  sycl::vec<T, 1> va{a};
+  using Tint =
```
Comment (yihanwg) [inline sycl/include/syclcompat/math.hpp:157]:
```suggestion
  using ByteT =
```
I think `ByteT` might be better, WDYT?

Comment (OuadiElfarouki) [reply sycl/include/syclcompat/math.hpp:157]: Yeah was thinking about naming it after "type of int" as singed/unsigned. But sure I'll go with something like "IntT"

File: sycl/include/syclcompat/math.hpp [hunk 5]
```diff
@@ -177,6 +190,36 @@ inline constexpr RetT extend_vbinary2(AT a, BT b, RetT c,
   return sycl::vec<Tint, 2>{temp[0], temp[1]}.template as<sycl::vec<RetT, 1>>();
 }
 
+template <typename RetT, bool NeedSat, bool NeedAdd, typename AT, typename BT,
+          typename BinaryOperation>
+inline constexpr RetT extend_vbinary4(AT a, BT b, RetT c,
+                                      BinaryOperation binary_op) {
+  static_assert(std::is_integral_v<AT> && std::is_integral_v<BT> &&
```
Comment (yihanwg) [inline sycl/include/syclcompat/math.hpp:194]: I think we can simplify to check whether AT and BT is `int32_t`/`uint32_t`, WDYT?
```
template <class T>
constexpr inline bool is_i32_or_u32 = std::is_same_v<std::decay_t<T>, int32_t> || 
                                      std::is_same_v<std::decay_t<T>, uint32_t>;

```

File: sycl/include/syclcompat/math.hpp [hunk 6]
```diff
@@ -177,6 +190,36 @@ inline constexpr RetT extend_vbinary2(AT a, BT b, RetT c,
...
+      binary_op(extend_a[0], extend_b[0]), binary_op(extend_a[1], extend_b[1]),
+      binary_op(extend_a[2], extend_b[2]), binary_op(extend_a[3], extend_b[3])};
+  using Tint =
+      typename std::conditional<std::is_signed_v<RetT>, int8_t, uint8_t>::type;
+
+  if constexpr (NeedSat) {
+    int16_t min_val = 0, max_val = 0;
+    min_val = std::numeric_limits<Tint>::min();
+    max_val = std::numeric_limits<Tint>::max();
+    temp = detail::clamp(temp, {min_val, min_val, min_val, min_val},
```
Comment (yihanwg) [inline sycl/include/syclcompat/math.hpp:208]: We can use the following code to construct a 4 element `min_val`/`max_val` vector.
https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#_vec_interface
`explicit constexpr vec(const DataT& arg);`

```
sycl::vec<int16_t, 4>(min_val), sycl::vec<int16_t, 4>(max_val)
```

File: sycl/include/syclcompat/math.hpp (hunk 5 above)
Comment (OuadiElfarouki) [reply sycl/include/syclcompat/math.hpp:194]: Yes that should look better @yihanwg. We don't want to put it inside the `detail` namespace though, any suggestion?

File: sycl/include/syclcompat/math.hpp (hunk 6 above)
Comment (OuadiElfarouki) [reply sycl/include/syclcompat/math.hpp:208]: Will apply for the v2 variant as well (just to keep it consistent, as the v2 extend got merged already).

**Review by joeatodd — APPROVED** [review]
LGTM

**Comment by OuadiElfarouki** [comment]
@intel/llvm-gatekeepers PR is ready! Thank you.

**Pull request merged**